Charles-Trees-Excercises.py

```diff
@@ -0,0 +1,116 @@
+#Trees Excercises
+
+#Tree Node
+class TreeNode: 
+    def __init__(self, left, right, value):
+        self.left = left
+        self.right = right
+        self.value = value
+
+#Tree Construction
+class Tree:
+    def __init__ (self, root):
+        self.root = root
+
+    def add_node (self, value):
+        node = TreeNode(
+            left = None,
+            right= None,
+            value = value
+        )
+        n = self.root 
+
+        def _add(node, n):
+ 
+            if node.value > n.value:
+                if n.right != None:
+                    _add(node, n.right)
+                else: 
+                    n.right = node 
+                    
+            elif node.value < n.value: 
+                if n.left != None:
+                    _add(node, n.left)
+                else: 
+                    n.left = node
+
+        _add(node, n)
+
+    def validate (self):
+
+        def _validation_legwork(n):
+            if n.left == None and n.right == None:
+                return True
+            else:
+                if n.left.value > n.value:
+                    return False
+
+                elif n.right.value < n.value: 
+                    return False
+
+                else:
+                    return _validation_legwork(n.left) and _validation_legwork(n.right) 
+                    
+        return _validation_legwork(self.root)
+
+
+    def search (self, searched_value):
+        
+        try: 
+            int(searched_value)
+        except: 
+            print("The provided arguments is not an intenger; this function can only take integers")
+            return None
+        
+        def _search (n, searched_value): 
+
+            if n == None:
+                return False
+            else:
+                if searched_value == n.value:
+                    return True
+
+                elif searched_value < n.value:
+                    return _search(n.left, searched_value)
+
+                
+                elif searched_value >= n.value:
+                    return _search(n.right, searched_value)
+
+                
+                else:
+                    return False
+                
```

Comment:
I think a better order of the checks would be as follows

```python
if searched_value < n.value:
    return _search(n.left, searched_value)
elif searched_value > n.value:
    return _search(n.right, searched_value)
else
    return True
```

Reply:
for the last condition, if its not less or not greater its gotta be equal. Either ways your solution will work just fine, this is just one other way of doing it